### ISSUE TITLE
Enhancing metadata with info required for programmatic traversal

### ISSUE THREAD

**Issue by yongrenjie** [Urban-Analytics-Technology-Platform/poppusher#81]
As I understand it, for **Belgium** there is currently:

- geometries at various levels (in decreasing size: region, province, administrative district, municipality, submunicipality, statistical sector)
- population data at _municipality_ level, the second-smallest division

This data and metadata is fine for a human to look through and use, but cannot be traversed in a programmatic way just yet.
In this issue I'll describe what's lacking, and make some suggestions for what to change to make it possible.

I expect the points raised here to be relevant to all countries.
Thus, although I'm contextualising them using the Belgian data, the changes will affect the overall metadata schema and should be applied to every country we provide in popgetter.

I'm not sure how helpful it is, but in case you wish to see what my work, I threw together some Python scripts. Note that of these scripts, only (1) will run with the code on popgetter main branch as (2) and (3) rely on local changes.

1. [Plot the population per municipality given the current format of the data.](https://gist.github.com/yongrenjie/5fbb2f200518a2c663d203e853a7a549#file-1_process_from_current-py) You will see that this script requires a few 'magic constants', which is what is preventing us from traversing the data programmatically.
2. [Restructure the data according to the changes I propose below](https://gist.github.com/yongrenjie/5fbb2f200518a2c663d203e853a7a549#file-2_restructure_belgium_data-py)
3. [Plot the population per municipality given the restructured data.](https://gist.github.com/yongrenjie/5fbb2f200518a2c663d203e853a7a549#file-3_process_from_restructured-py) This script doesn't need any 'magic constants' beyond the fixed filepaths for metadata parquets, and the column names of the deserialised structs. For the CLI library, we would essentially be porting this code to Rust.

## 1. The `SourceDataRelease` itself

The `SourceDataRelease` here

https://github.com/Urban-Analytics-Technology-Platform/popgetter/blob/e13d729926e9a37364a9a9147cf5adf0730c91ed/python/popgetter/assets/be/census_tables.py#L41-L54

is present in the popgetter library, but is not exported as an asset in Dagster.
This needs to be made available.

## 2. Metadata for geometry file

In the `GeoDataFrame` exported by Dagster, there are several columns. The most important of these are the IDs of each geographic subdivision:

 - `cd_sector` (statistical sector)
 - `cd_sub_munty` (submunicipality)
 - `cd_munty_refnis` (municipality)
 - `cd_dstr_refnis` (administrative district)
 - `cd_prov_refnis` (province)
 - `cd_region_refnis` (region)

and their names. Each geography level has three columns for the names in Dutch, French, and German.

 - `tx_sector_descr_{nl,fr,de}`
 - `tx_sub_munty_{nl,fr,de}`
 - `tx_munty_descr_{nl,fr,de}`
 - `tx_adm_dstr_descr_{nl,fr,de}`
 - `tx_prov_descr_{nl,fr,de}`
 - `tx_rgn_descr_{nl,fr,de}`

The `SourceDataRelease` tells us that the population data is provided at municipality level.
However, when the programme reads this in, it cannot identify which columns of the `GeoDataFrame` it should look up the municipality IDs and names from.

The geometries must thus be enriched with some kind of metadata which _looks like_:

```json
[
  {
    "level": "municipality",
    "id": "cd_munty_refnis",
    "name": ["tx_munty_descr_nl", "tx_munty_descr_fr", "tx_munty_descr_de"]
  },
  {
    "level": "administrative_district",
    "id": "cd_dstr_refnis",
    "name": ["tx_adm_dstr_descr_nl", "tx_adm_dstr_descr_fr", "tx_adm_dstr_descr_de"]
  },
  (...)
]
```

That way when the programme sees `municipality` in the `SourceDataRelease` it knows that the IDs can be found in `cd_munty_refnis`, and that there are three names for it in those respective columns.

Going one step further: to _tie_ this metadata to the geometries themselves, what we could do is to define a new type for geometry data:

```python
class GeographyRelease(BaseModel):
    name: str
    date_published: datetime.date
    # ^ Plus any other info we want to include...
    filepath: str   # could be multiple paths for different output formats

class GeographyReleaseLevel(BaseModel):
    release_name: str        # name of the above
    level_name: str          # "municipality"
    code_column: str         # "cd_munty_refnis"
    name_columns: list[str]  # ["tx_munty_descr_" + s for s in ["nl", "fr", "de"]]
```

Then the `SourceDataRelease`, instead of referring to a `geography_file` and `geography_level`, would refer to a `geography_release_name` and `geography_level_name`.

### 3. All nested types should be stored as IDs OR primary keys instead of the types themselves

This includes
 - `SourceDataRelease.publishing_organisation`
 - `SourceDataRelease.countries_of_interest`
 - `DataPublisher.countries_of_interest`
 - possibly anything else I missed

There are two reasons for this.
1. Storing them as the types themselves leads to duplication of information. This means:
   - querying will require the transmission of more data than needed
   - having the same object defined multiple times this can possibly lead to data inconsistency
2. They can't be serialised as parquet files 😢  `df.to_parquet()` will throw an error saying 'did not recognize Python value type when inferring an Arrow data type'.

To fix this, we need to assign each class instance a unique ID **or** ensure that they have a unique field, and store the value of that instead of the full type itself.

My personal inclination is that instead of trying to generate a unique ID for each `SourceDataRelease` (etc), it is easier to use the name as a primary key and look them up by name:
 - instead of storing `source_data_release_id` we store `source_data_release_name`
 - we have to ensure that `source_data_release_name` is unique between all `SourceDataRelease`s

This is, in my opinion, better than using IDs because generating unique, reproducible IDs is quite nontrivial:
 - `uuid4()` gives different results each time the pipeline is run and so is not reproducible. Say for example someone uses the CLI to fetch a metric with a given UUID — this is a very natural operation to do if the user already knows what they want. But if we rerun the pipeline or update the data then that UUID will be invalidated and their workflow will break.
 - if we use a counter as the ID we have to make sure that there are no clashes.
   This is OK if we are inserting into a database because we can set a column to autoincrement, but in Python we would have to use some kind of global variable,
   and the final result would depend on the order of execution as well as which assets are materialised, and that's very ugly.
 - if we construct some kind of hash from the fields, a unique hash necessitates having unique field(s), so why not just use the unique field as the key instead.

### 4. Data tables need to have the geoid column set as the index

This isn't an issue because it's already true for Belgium, but it's something that we need to be cognisant of.
It can't just be an ordinary column in the dataframe because then the CLI will need to know which column to pick up the IDs from.

----------------------

# Overall, what does each country need to export?

If we go with the suggestions above, then each country's pipeline should output the following:

 - (dataframe) **A list of `GeographyRelease`s**

   There can be multiple of these because we may have, say, 2011 and 2021 census data for the UK where the boundaries are different.

   - (various output formats) **Geometry files**

     Each `GeographyRelease` will point to the file(s) containing the geometries.

 - (dataframe) **A list of `GeographyReleaseLevel`s**

 - (dataframe) **A list of `MetricMetadata`s**

   This would enumerate all data that are available and tell the programme where to find them.

   - (parquet) **Data files**

     Each `MetricMetadata` will point to the file containing the data (in parquet format).

 - (dataframe) **A list of `SourceDataRelease`s**

 - (dataframe) **A list of `DataPublisher`s**

 - (dataframe) **A `CountryMetadata`**

Because the five dataframes all refer to each other, I do feel like it is _natural_ to store them in a relational database.
In this case, a single database will need to be exported with a standardised filename, and the table names should also be standardised.
_However_, as Stuart pointed out, this will likely cause issues compiling to WASM:
 - it seems the maintainers of the `duckdb-rs` library have no intention of supporting WASM (https://github.com/duckdb/duckdb-rs/issues/56)
 - rusqlite is more interested in doing it, and there's an PR for it, but hasn't been merged (https://github.com/rusqlite/rusqlite/pull/1010)

**So for now, I'd suggest that we export them as separate (parquet?) files.**
In this case they will all need to have predefined and standardised filenames.
Alternatively, we could have one single top-level file for each country (say `be/info.json`) which points to the five files.

---------------

# Going forward

I'd really appreciate any comments or suggestions on the above.
I think the _issues_ that I've pointed out are not subjective --- they're things that we _will_ need to fix.
However, it's very conceivable that there may be better ways to fix them than what I've suggested above!
I've tried to be opinionated on this just so that we have something concrete to talk about.

Once we reach a consensus, I'm totally happy to make necessary changes on this repo as well.
(I have already done some of them locally so that my script can run.)

**Comment by stuartlynn** [comment]
Hey @yongrenjie thanks for putting this together. I had some thoughts on a few of these questions 

## The structure of geography files and linking them to metrics 

One of the challenges we will have with the data is that we will have the same metric  (Say number of people between the age of 10 and 20) potentially published on each of the geographic levels for a country. So for example, we might it published on the block group level and then at the county level. These numbers might not necessarily add up to each (if we sum up all the values in the block groups contained in a county, we might not get the same value as reported at the county level) other because of privacy protecting strategies or other estimation methods in the census. So we want to host the data independently for each geographic level.

This means that we would have a SourceDataRelease for each geographic level. So there would be one entry for US blocks and another for US tracts and another for US counties etc. Each of these SourceDataReleases should point to a different geography file and metrics files which should be linked by a GeoID.

What this means is I think we should have 

- Separate geography files for each level / year in a countries census 
- Those files being a FGB with just the geography and a column called "geo_id" or something similar 
- A parquet files (or set of parquet files depending on how large the number of metrics is) for each geography which contains a column called "geo_id" which can be used to join with the FGB at that level and then columns for each of the metrics.

I think what we need to change with the Belgium data is not then the meta data structure but the output format of the resulting dataframes.

## Making the metadata relational rather than nested.

The other comments you make I think are mostly about making the metadata more relational rather than nested. I think there are pros and cons for each. The nested data might be more useful for a catalogue format while the relational version might be more useful for search and querying. 

I could go either way and have a slight preference for the relational structure you propose here but we could also write some utility functions to convert from the nested format to the relational when the data is loaded in to the tool. This might be a good way to go as it separates the external representation from the internal operational representation. 

We will have to do this anyway if we decide to go with an in memory DB in rust. This DB will likely be relational so we will need to write some code to load it in. 

That might let us leave a more human readable format of the metadata for other use cases. 

I propose we use tomorrows meeting to talk more about this and get us all aligned.

**Comment by yongrenjie** [comment]
> I think what we need to change with the Belgium data is not then the meta data structure but the output format of the resulting dataframes.

That sounds fine to me from the perspective of navigating the data! The only downside I see is that this would mean re-hosting what's effectively the same geometry multiple times. If we don't think this will be a problem in practice (i.e. the number of files we're hosting is limited) then I'm all for it.

-------

> The nested data might be more useful for a catalogue format while the relational version might be more useful for search and querying.

The way I see it, our internal representation of the data should be relational (for the reasons above - avoiding duplication and enabling serialisation), and the functions we use to query it can return nested data. (Or to be precise, we could say that the API the library exposes will return nested data)

> I propose we use tomorrows meeting to talk more about this and get us all aligned.

Yes, let's! I'll try and prepare something to talk through at the meeting so that I'm not just reading from this issue.

**Comment by stuartlynn** [comment]
> > I think what we need to change with the Belgium data is not then the meta data structure but the output format of the resulting dataframes.
> 
> That sounds fine to me from the perspective of navigating the data! The only downside I see is that this would mean re-hosting what's effectively the same geometry multiple times. If we don't think this will be a problem in practice (i.e. the number of files we're hosting is limited) then I'm all for it.

I dont think this means hosting the same geometry multiple times. We would have a single row for each geography in the FGB files, representing it only once and then a single row corresponding to the metrics associated with that geography in a parquet file.

Lets say we have a hypothetical country  with the following setup:

- The country has 3 counties, county A, B and C.
- Each county has 2 subregions called output areas, designated A_OA1,  A_OA2, B_OA1,B_OA2 etc
- The country publishes two metrics M1 and M2 at the output area level and at the county level.

So the resulting files would be 

- A FGB file containing the county geographies. It would have one entry each for county A, B and C representing the boundary geography for each along with a geo_id for each 
- A FGB file containing the output area geographies. It would have 6 entries representing the geographies of the output areas A_OA1, A_OA2, B_OA1, B_OA2, C_OA1, C_OA2 and the geo_ids for each 
- A parquet file with the metrics for the counties which would contain a row per county and a column per metric + 1 column for the linking geo_id  so 3 (counties)x 2(metrics) + 1 (geo_id column)
- A parquet file with the metrics for the output areas which would contain a row per county and a column per metric + 1 column for the linking geo_id  so 6(output areas) x 2(metrics) + 1 (geo_id column)

The geometries are only represented once each and can be linked to the respective parquet files to retrieve the required metrics.

The meta data then, for a given release at a given geometry level for a given year just needs to specify which geography file it refers to and which parquet files contain the metrics for those geographies.

**Comment by stuartlynn** [comment]
> The way I see it, our internal representation of the data should be relational (for the reasons above - avoiding duplication and enabling serialisation), and the functions we use to query it can return nested data. (Or to be precise, we could say that the API the library exposes will return nested data)

Yeah I could get behind that.

**Comment by yongrenjie** [comment]
<img width="859" alt="popgetter_diagram" src="https://github.com/Urban-Analytics-Technology-Platform/popgetter/assets/22414895/ee82f724-e61b-402f-a765-e26c44e96c15">

Here's a cleaned-up diagram of all the moving parts here.

Remaining questions (black text, highlighted in diagram) include:

1. What remaining fields are needed in each struct?
   - most of this is already present in the existing codebase and just needs some small tweaks to account for the changes we're proposing
2. Identifiers for each metadata struct: how do we generate them?
   - we agreed to have these as strings which are formed by combining the smallest set of fields which make the struct unique — so we need to identify which fields these are
3. How do we tie the two geography files (geometries + names) together?
   - this doesn't need to be a complicated decision, it can be as simple as having their filename stems be the same with a different extension
4. How do we relate geographies at different levels to each other?

I'd like to propose that we deal with (1) in this issue, and I'll open separate issues for (2) through (4).

**Comment by yongrenjie** [comment]
Update: having looked quite thoroughly through the current fields, I don't think much actually needs to be changed, just the descriptions need to be updated to match what we discussed yesterday.

I'll put in a PR for this!

**Comment by sgreenbury** [comment]
Following discussion with @yongrenjie and @andrewphilipsmith regarding generality of subdivision codes (i.e. currently using ISO3116-2 but other options such as [P-codes](https://humanitarian.atlassian.net/wiki/spaces/imtoolbox/pages/222265609/P-codes+and+gazetteers), opening #85 for further development of this.
